:seedling: Add maketarget for ensuring docker is present for testing

sirpaleet opened the pull request:
**What this PR does / why we need it**:

Some people may find the installation process and running the hack e2e-scripts a bit difficult, so making a target for the script and fixing the installed packages and dependencies should be done to make it easier to run them on a VM, and documentation should be added as well.

There is some previous discussion in here: #2642 

Fixes: #2707

tuminoid commented on Makefile:
```diff
@@ -120,6 +123,58 @@ unit-verbose: ## Run unit tests with verbose output
+## These installations work for linux amd64, at least on Ubuntu 24.04
+## NOTE: you may need to log out and back in to apply all changes
+.PHONY: e2e-prerequisites
+e2e-prerequisites:
```
I don't think we want this much bash in Makefile, rather have a new setup script that the makefile runs.

nuhakala replied:
Adding checks for existing installations will be also easier in a bash script.

nuhakala commented on Makefile:
```diff
@@ -120,6 +123,58 @@ unit-verbose: ## Run unit tests with verbose output
 
 ARTIFACTS ?= ${ROOT_DIR}/test/e2e/_artifacts
 
+## Additional checks for if installations exist could be added
+## The installations could be made to the project's bin, but
+## then in other places the scripts would not know to find them
+##
+## NEXT:
+## - `test-e2e` is used in ci-e2e.sh script so make a new target
+## - expand on `clean-e2e` to have optional steps on cleanup with flags
+## - update info in docs/testing.md and test/e2e/README.md
+##
+## Installing required packages for running e2e scripts locally
+## These installations work for linux amd64, at least on Ubuntu 24.04
+## NOTE: you may need to log out and back in to apply all changes
```
I think this comment could be rewritten to be slightly more understandable. The first paragraph seems to have two points, maybe create a list or two paragraphs from it. The NEXT section seems to be like things to improve next, maybe it could be renamed as TODO or "things to improve next". Now I got the impression that I need to do those things next after running this target (though the context reveals that this is not the case).

Also "but then in other places the scripts would not know to find them" -> "the scripts would not know how to find them" ?

sirpaleet replied:
Ah this was supposed to be notes for this PR but I'll take them off and make a separate comment on here then :D

sirpaleet commented:
Additional checks for if installations exist could be added.
The installations could be made to the project's bin, but then in other places the scripts would not know how to find them.

TODO:
- `test-e2e` is used in ci-e2e.sh script so make a new target
- expand on `clean-e2e` to have optional steps on cleanup with flags
- update info in docs/testing.md and test/e2e/README.md

sirpaleet commented on hack/e2e-prerequisites.sh:
```diff
@@ -0,0 +1,55 @@
+echo "INSTALLING some core packages"
+sudo apt-get install build-essential tar
+echo "INSTALLING go, removing old versions"
+sudo rm -rf /usr/local/go
```
Ah this was from the official instructions, I'll take it off then

tuminoid replied:
For example, on my machine, it would nuke all of my local Go versions, not just one. Go doesn't care where you put it, as long as the go environment variables and $PATH is correct.

sirpaleet replied:
I added the comment on this being run on vm:s only, I'll use these then :)

sirpaleet commented on hack/e2e-prerequisites.sh:
```diff
@@ -0,0 +1,55 @@
+echo "INSTALLING go, removing old versions"
+sudo rm -rf /usr/local/go
+wget -P $HOME https://go.dev/dl/go$GO_VERSION.linux-amd64.tar.gz
+sudo tar -C /usr/local -xzf $HOME/go$GO_VERSION.linux-amd64.tar.gz
```
Is this a general convention or somehow related to securing the read of the path etc?

tuminoid commented on hack/e2e-prerequisites.sh:
```diff
@@ -0,0 +1,55 @@
+# -----------------------------------------------------------------------------
+
+USR_LOCAL_BIN="/usr/local/bin"
+MINIMUM_KUBECTL_VERSION=v1.28.1
```
Maybe make another PR where you bump the kubectl version in ensure_kubectl.sh to 1.33.x

sirpaleet commented:
This has now been confined to the scope of installations, and adding the script target. I'll open the PR but I will most likely not have the time to edit all changes if they are major. I'll make an issue on the "TODO" i added before.

> I noticed that in here: https://github.com/metal3-io/baremetal-operator/blob/main/hack/ci-e2e.sh#L57 there is clear attempt at making the script self-sufficient and it should probably install all needed packages. But alas it does not. I'll expand on just this and remove the new target :D + edit the other scripts to fit this

Additionally based on this finding, changed from a new target for installations to adding the necessary steps to the hack/ci-e2e.sh and making a new ensure_docker.sh for the docker installation.

sirpaleet commented on hack/e2e/ensure_docker.sh:
```diff
@@ -0,0 +1,32 @@
+            # shellcheck source=/etc/os-release
+            echo \
+              "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \
+              $(. /etc/os-release && echo "${UBUNTU_CODENAME:-$VERSION_CODENAME}") stable" | \
```
Not quite sure how to deal with this shellcheck waring now, will look at in morning if I have time. This did not work here.

sirpaleet commented on hack/e2e/ensure_docker.sh:
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+# shellcheck disable=SC1091
```
Added this to fix the shellcheck issue, this code is from the official installation instructions of installing Docker Engine on Ubuntu.

tuminoid commented on Makefile:
```diff
@@ -120,6 +120,15 @@ unit-verbose: ## Run unit tests with verbose output
+## To note: Run this on your VM and not your local machine. These installations
+## work for linux amd64, at least on Ubuntu 24.04. You may need to log out and
+## log in to apply all changes if the installations are done for the first time.
+.PHONY: run-e2e-script
```
this is not descriptive name for the target... setup-e2e or such would be nicer

sirpaleet commented on hack/e2e/ensure_docker.sh:
```diff
@@ -0,0 +1,33 @@
+            sudo apt-get update
+            sudo apt-get install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
+            sudo usermod -aG docker "${USER}"
+            newgrp docker
```
I'm confused as to which part, since it works for me? The new docker group?

sirpaleet commented:
Left a couple of comments, and fixed some things. This is as much as I have time today, in case someone has an opinion of the furture of this PR I'm all ears but contributing to this has to end on my part for now :( I'll try to answer if someone has a question though!

terror96 commented on hack/e2e/ensure_go.sh:
```diff
@@ -1,34 +1,55 @@
+                -C /usr/local \
+                -xzf "/tmp/${MINIMUM_GO_VERSION}.linux-amd64.tar.gz"
+            export PATH="${PATH}:/usr/local/go/bin"
+            GO="$(command -v go)"
```
We are missing `set +x` here. I'll add it...

tuminoid commented on hack/e2e/ensure_go.sh:
```diff
@@ -1,34 +1,56 @@
+
 # Ensure the go tool exists and is a viable version, or installs it
 verify_go_version()
 {
```
```suggestion
verify_go_version() {
```
Might as well clean this up.

tuminoid reviewed (commented):
Some nits.

terror96 commented on hack/e2e/ensure_docker.sh:
```diff
@@ -0,0 +1,137 @@
+    DOCKER="$(command -v docker || true)"
+    if ! [[ -x "${DOCKER}" ]]; then
+        if [[ "${VERIFY_ONLY}" != "false" ]]; then
+            echo "docker is not installed or not in PATH"
```
Okay, reverting...
